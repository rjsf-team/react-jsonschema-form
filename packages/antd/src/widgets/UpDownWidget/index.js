import React from 'react';

import InputNumber from 'antd/lib/input-number';

const INPUT_STYLE = {
  width: '100%',
};

const UpDownWidget = ({
  // autofocus,
  disabled,
  formContext,
  id,
  onBlur,
  onChange,
  onFocus,
  // options,
  placeholder,
  readonly,
  // required,
  // schema,
  value,
}) => {
  const { readonlyAsDisabled = true } = formContext;

  const handleChange = (nextValue) => onChange(nextValue);

  const handleBlur = ({ target }) => onBlur(id, target.value);

  const handleFocus = ({ target }) => onFocus(id, target.value);

  return (
    <InputNumber
      disabled={disabled || (readonlyAsDisabled && readonly)}
      id={id}
      name={id}
      onBlur={!readonly ? handleBlur : undefined}
      onChange={!readonly ? handleChange : undefined}
      onFocus={!readonly ? handleFocus : undefined}
      placeholder={placeholder}
      style={INPUT_STYLE}
      type="number"
      value={value}
    />
  );
};

export default UpDownWidget;
