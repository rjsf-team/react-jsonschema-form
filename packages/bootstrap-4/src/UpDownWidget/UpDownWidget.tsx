import React from "react";

import Form from "react-bootstrap/Form";

import { WidgetProps } from "@rjsf/core";

const UpDownWidget = ({
  id,
  required,
  readonly,
  disabled,
  label,
  value,
  onChange,
  onBlur,
  onFocus,
  autofocus,
}: WidgetProps) => {
  const _onChange = ({
    target: { value },
  }: React.ChangeEvent<HTMLInputElement>) => onChange(value);
  const _onBlur = ({ target: { value } }: React.FocusEvent<HTMLInputElement>) =>
    onBlur(id, value);
  const _onFocus = ({
    target: { value },
  }: React.FocusEvent<HTMLInputElement>) => onFocus(id, value);

  return (
    <Form.Group  className="mb-0">
      <Form.Label>
        {label}
        {label && required ? "*" : null}
      </Form.Label>
      <Form.Control
        id={id}
        autoFocus={autofocus}
        required={required}
        type="number"
        disabled={disabled}
        readOnly={readonly}
        value={value || value === 0 ? value : ""}
        onChange={_onChange}
        onBlur={_onBlur}
        onFocus={_onFocus}
      />
    </Form.Group>
  );
};

export default UpDownWidget;
