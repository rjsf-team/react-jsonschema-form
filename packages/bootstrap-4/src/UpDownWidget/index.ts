export { default } from './UpDownWidget';
export * from './UpDownWidget';
