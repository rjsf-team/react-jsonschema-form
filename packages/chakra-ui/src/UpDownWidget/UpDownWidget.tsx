import { FocusEvent } from 'react';
import {
  NumberInput,
  NumberDecrementStepper,
  NumberIncrementStepper,
  NumberInputField,
  NumberInputStepper,
  FormControl,
  FormLabel,
} from '@chakra-ui/react';
import {
  ariaDescribedByIds,
  labelValue,
  FormContextType,
  RJSFSchema,
  StrictRJSFSchema,
  WidgetProps,
} from '@rjsf/utils';
import { getChakra } from '../utils';

export default function UpDownWidget<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: WidgetProps<T, S, F>
) {
  const { id, uiSchema, readonly, disabled, label, hideLabel, value, onChange, onBlur, onFocus, rawErrors, required } =
    props;

  const chakraProps = getChakra({ uiSchema });

  const _onChange = (value: string | number) => onChange(value);
  const _onBlur = ({ target: { value } }: FocusEvent<HTMLInputElement | any>) => onBlur(id, value);
  const _onFocus = ({ target: { value } }: FocusEvent<HTMLInputElement | any>) => onFocus(id, value);

  return (
    <FormControl
      mb={1}
      {...chakraProps}
      isDisabled={disabled || readonly}
      isRequired={required}
      isReadOnly={readonly}
      isInvalid={rawErrors && rawErrors.length > 0}
    >
      {labelValue(<FormLabel htmlFor={id}>{label}</FormLabel>, hideLabel || !label)}
      <NumberInput
        value={value ?? ''}
        onChange={_onChange}
        onBlur={_onBlur}
        onFocus={_onFocus}
        aria-describedby={ariaDescribedByIds<T>(id)}
      >
        <NumberInputField id={id} name={id} />
        <NumberInputStepper>
          <NumberIncrementStepper />
          <NumberDecrementStepper />
        </NumberInputStepper>
      </NumberInput>
    </FormControl>
  );
}
