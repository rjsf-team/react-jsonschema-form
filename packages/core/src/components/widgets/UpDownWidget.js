import React from "react";
import PropTypes from "prop-types";

import { rangeSpec } from "@rjsf/utils";

function UpDownWidget(props) {
  const {
    registry: {
      widgets: { BaseInput },
    },
  } = props;
  return <BaseInput type="number" {...props} {...rangeSpec(props.schema)} />;
}

if (process.env.NODE_ENV !== "production") {
  UpDownWidget.propTypes = {
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  };
}

export default UpDownWidget;
