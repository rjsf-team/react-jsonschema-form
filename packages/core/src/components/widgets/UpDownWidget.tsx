import { getTemplate, FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';

/** The `UpDownWidget` component uses the `BaseInputTemplate` changing the type to `number`.
 *
 * @param props - The `WidgetProps` for this component
 */
export default function UpDownWidget<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: WidgetProps<T, S, F>
) {
  const { options, registry } = props;
  const BaseInputTemplate = getTemplate<'BaseInputTemplate', T, S, F>('BaseInputTemplate', registry, options);
  return <BaseInputTemplate type='number' {...props} />;
}
