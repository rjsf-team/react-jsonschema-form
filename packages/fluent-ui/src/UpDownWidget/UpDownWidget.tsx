import React from "react";
import { Label } from "@fluentui/react";
import { SpinButton } from "@fluentui/react";
import { WidgetProps } from "@rjsf/core";
import _pick from "lodash/pick";
import { utils } from "@rjsf/core";

const { rangeSpec } = utils;

// Keys of ISpinButtonProps from @fluentui/react
const allowedProps = [
  "ariaDescribedBy",
  "ariaLabel",
  "ariaPositionInSet",
  "ariaSetSize",
  "ariaValueNow",
  "ariaValueText",
  "className",
  "componentRef",
  "decrementButtonAriaLabel",
  "decrementButtonIcon",
  "defaultValue",
  "disabled",
  "downArrowButtonStyles",
  "getClassNames",
  "iconButtonProps",
  "iconProps",
  "incrementButtonAriaLabel",
  "incrementButtonIcon",
  "inputProps",
  "keytipProps",
  "label",
  "labelPosition",
  "max",
  "min",
  "onBlur",
  "onDecrement",
  "onFocus",
  "onIncrement",
  "onValidate",
  "precision",
  "step",
  "styles",
  "theme",
  "title",
  "upArrowButtonStyles",
  "value",
];

const UpDownWidget = ({
  id,
  required,
  readonly,
  disabled,
  label,
  value,
  onChange,
  onBlur,
  onFocus,
  options,
  schema,
}: // autofocus,
WidgetProps) => {
  const _onChange = ({
    target: { value },
  }: React.ChangeEvent<HTMLInputElement>) => onChange(Number(value));

  let { min, max, step } = rangeSpec(schema);
  if (min === undefined) {
    min = -1 * Infinity;
  }
  if (max === undefined) {
    max = Infinity;
  }
  if (step === undefined) {
    step = 1;
  }

  const _onIncrement = (value: string) => {
    if (Number(value) + step! <= max!) onChange(Number(value) + step!);
  };

  const _onDecrement = (value: string) => {
    if (Number(value) - step! >= min!) onChange(Number(value) - step!);
  };

  const _onBlur = ({ target: { value } }: React.FocusEvent<HTMLInputElement>) =>
    onBlur(id, value);
  const _onFocus = ({
    target: { value },
  }: React.FocusEvent<HTMLInputElement>) => onFocus(id, value);

  const requiredSymbol = required ? "*" : "";

  const uiProps = _pick(options.props || {}, allowedProps);

  return (
    <>
      {/* TODO: add label for= attribute */}
      <Label htmlFor={id}>{label + requiredSymbol}</Label>
      <SpinButton
        id={id}
        min={min}
        max={max}
        step={step}
        incrementButtonAriaLabel={"Increase value by 1"}
        decrementButtonAriaLabel={"Decrease value by 1"}
        disabled={disabled || readonly}
        value={value || value === 0 ? value : ""}
        onBlur={_onBlur}
        onFocus={_onFocus}
        onChange={_onChange}
        onIncrement={_onIncrement}
        onDecrement={_onDecrement}
        for=""
        {...uiProps}
      />
    </>
  );
};

export default UpDownWidget;
