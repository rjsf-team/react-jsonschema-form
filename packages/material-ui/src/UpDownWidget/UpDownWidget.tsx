import React from 'react';

import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';

import { WidgetProps } from '@rjsf/core';

const UpDownWidget = ({
  id,
  required,
  readonly,
  disabled,
  label,
  value,
  onChange,
  onBlur,
  onFocus,
  autofocus,
}: WidgetProps) => {
  const _onChange = ({
    target: { value },
  }: React.ChangeEvent<HTMLInputElement>) => onChange(value);
  const _onBlur = ({ target: { value } }: React.FocusEvent<HTMLInputElement>) =>
    onBlur(id, value);
  const _onFocus = ({
    target: { value },
  }: React.FocusEvent<HTMLInputElement>) => onFocus(id, value);

  return (
    <FormControl
      fullWidth={true}
      //error={!!rawErrors}
      required={required}
    >
      <InputLabel>{label}</InputLabel>
      <Input
        id={id}
        autoFocus={autofocus}
        required={required}
        type="number"
        disabled={disabled || readonly}
        value={value || value === 0 ? value : ''}
        onChange={_onChange}
        onBlur={_onBlur}
        onFocus={_onFocus}
      />
    </FormControl>
  );
};

export default UpDownWidget;
