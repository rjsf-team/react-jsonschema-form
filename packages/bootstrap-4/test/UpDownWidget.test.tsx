import React from "react";
import UpDownWidget from "../src/UpDownWidget";
import renderer from "react-test-renderer";
import { makeWidgetMockProps } from "./helpers/createMocks";

describe("UpDownWidget", () => {
  test("renders 0 as 0 and not ''", () => {
    const tree = renderer
      .create(<UpDownWidget {...{ ...makeWidgetMockProps({}), value: 0 }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
