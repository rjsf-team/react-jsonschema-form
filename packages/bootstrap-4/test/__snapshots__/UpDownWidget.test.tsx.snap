// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpDownWidget renders 0 as 0 and not '' 1`] = `
<div
  className="mb-0 form-group"
>
  <label
    className="form-label"
  >
    Some simple label
    *
  </label>
  <input
    autoFocus={true}
    className="form-control"
    disabled={false}
    id="_id"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    readOnly={true}
    required={true}
    type="number"
    value={0}
  />
</div>
`;
